src/engine/Join.cpp

```diff
@@ -47,8 +47,8 @@ Join::Join(QueryExecutionContext* qec, std::shared_ptr<QueryExecutionTree> t1,
   // If one of the inputs is a SCAN and the other one is not, always make the
   // SCAN the right child (which also gives a deterministic order of the
   // subtrees). This simplifies several branches in the `computeResult` method.
-  if (t1->getType() == QueryExecutionTree::SCAN &&
-      t2->getType() != QueryExecutionTree::SCAN) {
+  if (std::dynamic_pointer_cast<IndexScan>(t1->getRootOperation()) &&
+      !std::dynamic_pointer_cast<IndexScan>(t2->getRootOperation())) {
```

Comment:
I think we can implement a templated member function for the `QueryExecutionTree` class to reduce the visual overhead and we can write something like

```
if (t1->as<IndexScan>() && ts->as<IndexScan>) {...}
```
If you want you can even add an `is` function that makes the code more readable (just compare the output of as with the nullptr.

I suppose that in most cases a `dynamic_cast` suffices. If not, you can still also add `asShared` which does the `dynamic_pointer_cast` which is a little bit more expensive because of the refcount increase.